specs/Floating-Point-Extra.md

```diff
@@ -0,0 +1,108 @@
+
+Floating Point Exceptions:
+- INVALID: If an operand to an instruction is outside of it's domain
+- OVERFLOW: If an operation produces a value that is too large for the destination format
+- UNDERFLOW: If an operation produces a value that is too small for the destination format
+- INEXACT: If an operation with a rational result produces a value that cannot be exactly represented in the destiniation format.
+- INEXACT: If the operation is performed with greater precision than the destination format, and the result cannot be exactly represented in the destiniation format
+- UNDERFLOW: If the result would be a subnormal value and fpcrw.DENORM is clear.
+
+
+Flags: If `f` is not set in `h`, `N`, `Z`, and `P` are set according to the result. 
+
+Instructions:
+- 0x100 (exp): Computes exp(x) of the operand, and stores the result in the operand.
+- 0x101 (ln): Computes ln(x) (the natural logarithm) of the operand, and stores the result in the operand
+- 0x102 (lg): Computes lg(x) (the base 2 logarithm) of the operand, and stores the result in the operand.
+- 0x103 (sin): Computes sin(x) of the operand, and stores the result in the operand.
+- 0x104 (cos): Computes cos(x) of the operand, and stores the result in the operand
+- 0x105 (tan): Computes tan(x) of the operand, and stores the result in the operand.
+- 0x106 (asin): Computes asin(x) of the operand, and stores the result in the operand.
+- 0x107 (acos): Computes acos(x) of the operand, and stores the result in the operand.
+- 0x108 (atan): Computes atan(x) of the operand, and stores the result in the operand.
+- 0x109 (sinh): Computes sinh(x) (the hyperbolic sin) of the operand, and stores the result in the operand.
+- 0x10a (cosh): Computes cosh(x) (the hyperbolic cos) of the operand, and stores the result in the operand.
+- 0x10b (tanh): Computes tanh(x) (the hyperbolic tan) of the operand, and stores the result in the operand.
+- 0x10c (asinh): Computes asinh(x) (the inverse hyperbolic sin) of the operand, and stores the result in the operand
+- 0x10d (acosh): Computes acosh(x) (the inverse hyperbolic cos) of the operand, and stores the result in the operand.
+- 0x10e (atanh): Computes atanh(x) (the inverse hyperbolic tan) of the operand, and stores the result in the operand.
+- 0x10f (exp2): Computes 2^x of the operand, and stores the result in the operand.
+- 0x113 (log10): Computes log10(x) (the base 10 logarithm) of the operand, and stores the result in the operand
+- 0x114 (lnp1): Computes ln(x+1) of the operand, and stores the result in the operand
+- 0x115 (expm1): Computes exp(x-1) of the operand, and stores the result in the operand.
+- 0x119 (sqrt): Computes sqrt(x) of the operand, and stores the result in the operand.
+- 0x11a (cbrt): Computes cbrt(x) of the operand, and stores the result in the operand.
+- 0x125 (hypot): Computes the hypotenuse of the two operands and stores the result in the first.
```

Comment:
What is the "hypotenuse" of two operands?

Reply:
sqrt(op1^2+op2^2) w/o intermediate loss of precision.

Reply:
This is definitely redundant and expensive. You need an entirely separate unit for this that essentially just has copies of the sqrt module and squaring module, but with higher precision. Possibly significantly higher.

Reply:
The problem is the same with fma - the software implementation w/o loss of precision is difficult if not nearly impossible.

Reply:
Just read the glibc code; it doesn't actually do that much to correct for error. It doesn't actually need to do much.

Reply:
(meanwhile, the i386 impl just uses 80-bit from the get-go)

Reply:
(... really, so does the amd64 impl)

Reply:
(... meanwhile, LLVM's implementation looks hectic, to say the least)

Reply:
Keep in mind that clever only has up to 64-bit (and may get 128-bit, but never 80-bit extended precision).